### PR TITLE
Add docs for `fsdp_native`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Added docs for `fsdp_native`.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @akihironitta

### CODE REVIEW

**Review by akihironitta — APPROVED** [review]

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 1]
```diff
@@ -206,6 +206,81 @@ This saves memory when training larger models however requires wrapping modules
...
+
+.. code-block:: python
+
+    from pytorch_lightning import Trainer
+    from pytorch_lightning.strategies import DDPFullyShardedNativeStrategy
+    from torch.distributed.fsdp.fully_sharded_data_parallel import CPUOffload
+
+
+    native_fsdp = DDPFullyShardedNativeStrategy(cpu_offload=CPUOffload(offload_params=True))
+    trainer = pl.Trainer(strategy=native_fsdp, accelerator="gpu", device=s4)
```
Comment (kaushikb11) [inline docs/source-pytorch/advanced/model_parallel.rst:278]:
```suggestion
    trainer = pl.Trainer(strategy=native_fsdp, accelerator="gpu", device=4)
```

**Review by kaushikb11 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@Felonious-Spellfire ^^ pls :)

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 2]
```diff
@@ -224,6 +225,8 @@ def model_to_device(self) -> None:
 
     @contextlib.contextmanager
     def model_sharded_context(self) -> Generator:
+        # TODO: users should be able to initialize without wrapping their layers
+        # as shown in the example here: https://fairscale.readthedocs.io/en/stable/api/nn/fsdp.html
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:229]: Why add this TODO here? If relevant, create an issue for it.

Also this references fairscale instead of PyTorch

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 2 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:229]: removing this. Will create an issue, I added this to my personal backlog to investigate this first before creating an issue with POC.

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 3]
```diff
@@ -206,6 +206,81 @@ This saves memory when training larger models however requires wrapping modules
...
+        def configure_optimizers(self):
+            return torch.optim.AdamW(self.model.parameters())
+
+
+    model = MyModel()
+    trainer = Trainer(accelerator="gpu", devices=4, strategy="fsdp_native", precision=16)
+    trainer.fit(model)
+
+
+You can customize the strategy configuration by adjusting the arguments of :class:`~pytorch_lightning.strategies.fully_sharded_native.DDPFullyShardedNativeStrategy` and pass that to the ``strategy`` argumen inside the ``Trainer``.
```
Comment (awaelchli) [inline docs/source-pytorch/advanced/model_parallel.rst:268]:
```suggestion
You can customize the strategy configuration by adjusting the arguments of :class:`~pytorch_lightning.strategies.fully_sharded_native.DDPFullyShardedNativeStrategy` and pass that to the ``strategy`` argument inside the ``Trainer``.
```

**Review by awaelchli — APPROVED** [review]
